google-cloud-firestore/src/main/java/com/google/cloud/firestore/RecursiveDelete.java

```
@@ -0,0 +1,353 @@
+    this.firestoreRpcContext = firestoreRpcContext;
+    this.writer = writer;
+    this.collectionReference = reference;
+    this.documentReference = null;
+  }
+
+  RecursiveDelete(
+      FirestoreImpl firestoreRpcContext, BulkWriter writer, DocumentReference reference) {
+    Preconditions.checkState(reference != null, "DocumentReference cannot be null");
+    this.firestoreRpcContext = firestoreRpcContext;
+    this.writer = writer;
+    this.documentReference = reference;
+    this.collectionReference = null;
+  }
+
+  public ApiFuture<Void> run() {
+    Preconditions.checkState(
+        documentsPending, "The recursive delete operation has already been completed");
+
+    // TODO: figure out if this is safe, or a way to lock this.
+    writer.verifyNotClosedLocked();
+
+    streamDescendants();
+    return completionFuture;
+  }
+
+  private void streamDescendants() {
+    Preconditions.checkState(
+        (documentReference == null && collectionReference != null)
+            || (documentReference != null && collectionReference == null),
+        "Either documentReference or collectionReference should be set");
+
+    Query query;
+    if (collectionReference != null) {
+      query = getAllDescendantsQuery(collectionReference);
+    } else {
+      query = getAllDescendantsQuery(documentReference);
+    }
+    streamInProgress = true;
+    final int[] streamedDocsCount = {0};
+    final ApiStreamObserver<DocumentSnapshot> responseObserver =
+        new ApiStreamObserver<DocumentSnapshot>() {
+          public void onNext(DocumentSnapshot snapshot) {
+            streamedDocsCount[0]++;
+            lastDocumentSnap = snapshot;
+            deleteReference(snapshot.getReference());
+          }
+
+          public void onError(Throwable throwable) {
+            lastError =
+                new FirestoreException("Failed to fetch children documents.", Status.UNAVAILABLE);
```

This swallows the underlying exception.

---

Wrapped the old exception. The status should be `UNAVAILABLE` regardless of the error, as it is on Node.